src/Console/BaseCommand.php

```diff
@@ -164,11 +166,66 @@ public function initialize(): void
     {
     }
 
+    /**
+     * Returns a list of all events that will fire in the command during its lifecycle.
+     * You can override this function to add your own listener callbacks
+     *
+     * @return array<string, mixed>
+     */
+    public function implementedEvents(): array
+    {
+        return [
+            'Command.initialize' => 'beforeFilter',
+            'Command.beforeExecute' => 'beforeExecute',
+            'Command.afterExecute' => 'afterExecute',
+        ];
+    }
+
+    /**
+     * Called before the command argument parsing logic. You can use this method to configure and customize the command
```

Comment:
Can you give an example of when would need to use `Command.initialize` instead of `Command.beforeExecute`.

It is run before the arguments are passed before unless one can return modified arguments from the event's callback (again not sure for what use case), I don't really see it's purpose.

Reply:
I was thinking about tackling modified arguments in a follow-up PR, since IMO it's out of scope of this one.

Some use cases I can think of off-hand are:
- argument expansion that OptionParser doesn't support
- deprecating an old invocation and you want to shim it to the new one
- path normalization